kv: listen for quiesce signal in IncrementValRetryable

yuzefovich opened the pull request:
This commit makes it so that we specify `Closer` retry option in `IncrementValRetryable` to listen for the quiesce signal as well as updates the retry struct to also use the provided context. We just saw a test failure that occurred due to a slow quiesce of the test server.

Fixes: #120499.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120655)
<!-- Reviewable:end -->

yuzefovich commented:
This code hasn't really been touched in like 7 years, yet this seems like a clear omission, right? I added the backport label since we got the failure on 23.2, but let me know if you see any risk with this change.

yuzefovich commented:
Actually I think that we also want to specify the Closer option. Let me update the PR first.

nvb reviewed (commented):
Nice find!

> Reviewed 2 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120655#-:-NtI7qqB4i_Czbjw53Rc:bdkgtm2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/kv/db.go` line 1183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120655#-NtI7fkG-XAJ0XGezxZM:-NtI7fkH9zz4UaNAyec3:b-l920tk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e14925585cf9cbb4d1a81f121c635a91e888d8b4/pkg/kv/db.go#L1183)):*
> ```Go
> 	var res KeyValue
> 	retryOpts := base.DefaultRetryOptions()
> 	retryOpts.Closer = stopper.ShouldQuiesce()
> ```

Can we use `db.Context().Stopper.ShouldQuiesce()` here, instead of plumbing in a separate stopper?


<!-- Sent from Reviewable.io -->

yuzefovich committed:
kv: listen for quiesce signal in IncrementValRetryable

This commit makes it so that we specify `Closer` retry option in
`IncrementValRetryable` to listen for the quiesce signal as well as
updates the retry struct to also use the provided context. We just
saw a test failure that occurred due to a slow quiesce of the test
server.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120655)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/db.go` line 1183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120655#-NtI7fkG-XAJ0XGezxZM:-NtIDrWNDveIe5TTfDED:bblmapa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e14925585cf9cbb4d1a81f121c635a91e888d8b4/pkg/kv/db.go#L1183)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we use `db.Context().Stopper.ShouldQuiesce()` here, instead of plumbing in a separate stopper?
</blockquote></details>

Indeed, thanks for pointing this out, done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Friendly ping @nvanbenschoten - this should be a quick one.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 18 of 18 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120655#-:-Nu2HZ1PD34SNHnLgaUt:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

yuzefovich commented:
blathers backport release-24.1

rafiss commented:
Given that we saw this problem in the wild (https://github.com/cockroachlabs/support/issues/2957), should we backport this to 23.1 also? In 23.1, the usage of `retry.Start` rather than `retry.StartWithCtx` means that cancellations won't work in some cases.

blathers backport 23.1
